src/crypto/muhash.cpp

```diff
@@ -0,0 +1,304 @@
+    }
+}
+
+void Multiply(Num3072* in_out, const Num3072* a)
+{
+    limb_t c0 = 0, c1 = 0;
+    Num3072 tmp;
+
+    /* Compute limbs 0..N-2 of in_out*a into tmp, including one reduction. */
+    for (int j = 0; j < LIMBS - 1; ++j) {
+        limb_t d0 = 0, d1 = 0, d2 = 0, c2 = 0;
+        mul(d0, d1, in_out->limbs[1 + j], a->limbs[LIMBS + j - (1 + j)]);
+        for (int i = 2 + j; i < LIMBS; ++i) muladd3(d0, d1, d2, in_out->limbs[i], a->limbs[LIMBS + j - i]);
+        mulnadd3(c0, c1, c2, d0, d1, d2, MAX_PRIME_DIFF);
+        for (int i = 0; i < j + 1; ++i) muladd3(c0, c1, c2, in_out->limbs[i], a->limbs[j - i]);
+        extract3(c0, c1, c2, tmp.limbs[j]);
+    }
+    /* Compute limb N-1 of a*b into tmp. */
+    {
+        limb_t c2 = 0;
+        for (int i = 0; i < LIMBS; ++i) muladd3(c0, c1, c2, in_out->limbs[i], a->limbs[LIMBS - 1 - i]);
+        extract3(c0, c1, c2, tmp.limbs[LIMBS - 1]);
+    }
+    /* Perform a second reduction. */
+    muln2(c0, c1, MAX_PRIME_DIFF);
+    for (int j = 0; j < LIMBS; ++j) {
+        add2(c0, c1, tmp.limbs[j]);
+        extract2(c0, c1, in_out->limbs[j]);
+    }
+#ifdef DEBUG
+    assert(c1 == 0);
+    assert(c0 == 0 || c0 == 1);
+#endif
+    /* Perform a potential third reduction. */
+    if (c0) FullReduce(in_out);
+}
+
+void Square(Num3072* in_out)
+{
+    limb_t c0 = 0, c1 = 0;
+    Num3072 tmp;
+
+    /* Compute limbs 0..N-2 of in_out*in_out into tmp, including one reduction. */
+    for (int j = 0; j < LIMBS - 1; ++j) {
+        limb_t d0 = 0, d1 = 0, d2 = 0, c2 = 0;
+        for (int i = 0; i < (LIMBS - 1 - j) / 2; ++i) muldbladd3(d0, d1, d2, in_out->limbs[i + j + 1], in_out->limbs[LIMBS - 1 - i]);
+        if ((j + 1) & 1) muladd3(d0, d1, d2, in_out->limbs[(LIMBS - 1 - j) / 2 + j + 1], in_out->limbs[LIMBS - 1 - (LIMBS - 1 - j) / 2]);
+        mulnadd3(c0, c1, c2, d0, d1, d2, MAX_PRIME_DIFF);
+        for (int i = 0; i < (j + 1) / 2; ++i) muldbladd3(c0, c1, c2, in_out->limbs[i], in_out->limbs[j - i]);
+        if ((j + 1) & 1) muladd3(c0, c1, c2, in_out->limbs[(j + 1) / 2], in_out->limbs[j - (j + 1) / 2]);
+        extract3(c0, c1, c2, tmp.limbs[j]);
+    }
+    {
+        limb_t c2 = 0;
+        for (int i = 0; i < LIMBS / 2; ++i) muldbladd3(c0, c1, c2, in_out->limbs[i], in_out->limbs[LIMBS - 1 - i]);
+        extract3(c0, c1, c2, tmp.limbs[LIMBS - 1]);
+    }
+    /* Perform a second reduction. */
+    muln2(c0, c1, MAX_PRIME_DIFF);
+    for (int j = 0; j < LIMBS; ++j) {
+        add2(c0, c1, tmp.limbs[j]);
+        extract2(c0, c1, in_out->limbs[j]);
+    }
+#ifdef DEBUG
+    assert(c1 == 0);
+    assert(c0 == 0 || c0 == 1);
+#endif
+    /* Perform a potential third reduction. */
+    if (c0) FullReduce(in_out);
+}
+
+void Inverse(Num3072* out, const Num3072* a)
+{
+    // For fast exponentiation a sliding window exponentiation with repunit
+    // precomputation is utilized. See "Fast Point Decompression for Standard
+    // Elliptic Curves" (Brumley, Järvinen, 2008).
```

Comment:
I did spend quite some time looking for something but the paper is only available behind a paywall and a friend told me there are other sources but that we might not want to link to those. I didn't find any summaries or so on the topic so if we want more details I would need to write something up myself I think.

Reply:
This is just an explanation of how the constants were chosen in the first place. I think it's only interesting in the same way references in an academic paper help trace the lineage of an idea rather than actually useful for reviewing the code as it stands -- ie, handy if you want to reinvent it, eg if 3072 bits aren't enough and you need to pick a new prime and hence new constants for efficient inversion, or if you want to see if you can find different constants that would involve even fewer square/multiply operations to get the same result. But not needed if you just want to check correctness.

Reply:
Agree. I don't think there's anything more needed here.

Reply:
The paper if anyone wants: https://sci-hub.do/10.1007/978-3-540-69485-4_10
as for the technique AFAIU(didn't review the actual code here) it's simply fermat little theorem (a^p-2=1/a) together with a simple square-and-multiply algorithm (in elliptic curves it's called double-and-add) 
Good references:
https://en.wikipedia.org/wiki/Exponentiation_by_squaring
https://briansmith.org/ecc-inversion-addition-chains-01

Reply:
@elichai That's the easy part. It's specifically using a technique here to construct an efficient exponentiation ladder for an exponent with many consecutive 1s.

Reply:
@sipa you're right. I ignored how you came up with that specific ladder because I didn't actually review the ladder itself (ops)